### PR TITLE
Final logging updates

### LINKED ISSUE #2993: Composer Tracing - reboot
## Composer Logging
*What* and *Why* and *When* and *How* and *Where* and *Who*

### Why?
To be able to solve problems both as contributors and end-users.

### What?
Always hard to say exactly what to be logged in terms of the data. The code has different levels of logging that can be enabled. The origin of the log entries are at various points in the code. 

The standard NPM log levels of, Error, Warn, Info, Verbose, Debug, Silly are used. (Silly is used just to be log everything - we don't have any specific points at this level. Also a None level is used to signify nothing).

Individual methods are available on the `composer-common\Logger` class - along with `entry() exit()` for methods. These are logged at the debug level. 

### When?

When to log and at what level and components are controlled by the DEBUG control string, as used by many node applications. This is a comma-separated list of components. A single * means everything, and a - sign infront of any component means do *not* log this

```
DEBUG=<moduleA>,<moduleB>
```
 The string 'composer' is used to indetify Hyperledger Composer. For example

```
DEBUG=express,composer,http
```

Would log all the composer log points, as well as whatever Express and HTTP wanted to do.

#### when exactly?
To filter this at the earliest point, the syntax is extended to be

```
DEBUG=composer[tracelevel]:fqn/class/name
```
The [tracelevel] including the [ ] is optional and defaults to *error*;  the `fqn/class/name` is dependant on how the code is written and the name of the logger that it requests. (subject for more work later)

Examples are

- `composer:*`  Everything at *error* level
- `composer[error]:*`  Everything at *error* level
- `composer[info]:*` Everything at *info* level
- `composer[info]:BusinessNetworkConnection`  Solely *BusinessNetworkConnection* at the *info* level
- `-composer:BusinessNetworkConnection,composer[error]:*`  Do not trace anything in the BusinessNetworkConnection class, but do trace anywhere else that has errors.  *the do not takes effect at all levels of trace*

### LINKED ISSUE #3642: Adjust to cope with Fabric GA logging levels.
Related to #3637;  Hyperledger Fabric v1.1 GA appears to have different behavior in relation to the setting of the CORE_CHAINCODE_LOGGING_LEVEL environment variable.  Actual code changes to support this can't be located.

Namely all the chaincode containers start off at an INFO level of logging. The changes here make that understandable by Composer logging.

Also adding in integration tests for to validate the get/set of the log level from cli

### PR DESCRIPTION
**Pull request by mbwhite**
Signed-off-by: Matthew B White <whitemat@uk.ibm.com>

Closes #3642 #2993

### CODE REVIEW

**Comment by mbwhite** [comment]
image
Personal build

File: packages/composer-common/test/log/consolelogger.js [hunk 1]
```diff
@@ -63,47 +70,58 @@ describe('ConsoleLogger', () => {
...
+
+        afterEach(() => {
+            sandbox.restore();
+        });
 
         it('should log debug level', () => {
             const spy = sandbox.spy(console, 'log');
             logger.log('debug', 'func1', 'some msg', [ 'arg1', 'arg2' ]);
             sinon.assert.calledOnce(spy);
-            sinon.assert.calledWith(spy, 'func1 some msg arg1, arg2');
```
Comment (sstone1) [inline packages/composer-common/test/log/consolelogger.js]: y u remove assert?

File: packages/composer-common/test/log/consolelogger.js [hunk 2]
```diff
@@ -63,47 +70,58 @@ describe('ConsoleLogger', () => {
...
         });
 
         it('should log warn level', () => {
             const spy = sandbox.spy(console, 'warn');
             logger.log('warn', 'func1', 'some msg', [ 'arg1', 'arg2' ]);
             sinon.assert.calledOnce(spy);
-            sinon.assert.calledWith(spy, 'func1 some msg arg1, arg2');
```
Comment (sstone1) [inline packages/composer-common/test/log/consolelogger.js]: y u remove assert?

File: packages/composer-common/test/log/consolelogger.js [hunk 3]
```diff
@@ -63,47 +70,58 @@ describe('ConsoleLogger', () => {
...
+
         });
 
         it('should log info level', () => {
             const spy = sandbox.spy(console, 'info');
             logger.log('info', 'func1', 'some msg', [ 'arg1', 'arg2' ]);
             sinon.assert.calledOnce(spy);
-            sinon.assert.calledWith(spy, 'func1 some msg arg1, arg2');
```
Comment (sstone1) [inline packages/composer-common/test/log/consolelogger.js]: y u remove assert?

File: packages/composer-common/test/log/consolelogger.js [hunk 4]
```diff
@@ -63,47 +70,58 @@ describe('ConsoleLogger', () => {
...
+
         });
 
         it('should log verbose level', () => {
             const spy = sandbox.spy(console, 'log');
             logger.log('verbose', 'func1', 'some msg', [ 'arg1', 'arg2' ]);
             sinon.assert.calledOnce(spy);
-            sinon.assert.calledWith(spy, 'func1 some msg arg1, arg2');
```
Comment (sstone1) [inline packages/composer-common/test/log/consolelogger.js]: y u remove assert?

File: packages/composer-common/test/log/consolelogger.js [hunk 5]
```diff
@@ -63,47 +70,58 @@ describe('ConsoleLogger', () => {
...
+
         });
 
         it('should log error level', () => {
             const spy = sandbox.spy(console, 'error');
             logger.log('error', 'func1', 'some msg', [ 'arg1', 'arg2' ]);
             sinon.assert.calledOnce(spy);
-            sinon.assert.calledWith(spy, 'func1 some msg arg1, arg2');
```
Comment (sstone1) [inline packages/composer-common/test/log/consolelogger.js]: y u remove assert?

File: packages/composer-runtime/test/networkmanager.js [hunk 6]
```diff
@@ -151,16 +152,22 @@ describe('NetworkManager', () => {
...
+                'fakelogger':'config'
+            });
+            mockLoggingService.setLoggerCfg.resolves();
+            sinon.spy(Logger.setLoggerCfg);
             mockAccessController.check.resolves();
             mockRegistryManager.getAll.resolves([mockRegistry]);
             return networkManager.setLogLevel(mockApi, {newLogLevel:'level'})
                         .then(() => {
-                            sinon.assert.calledWith(mockLoggingService.setLogLevel,'level');
+                            sinon.assert.calledOnce(mockLoggingService.setLoggerCfg);
```
Comment (sstone1) [inline packages/composer-runtime/test/networkmanager.js]: y u remove assert?

File: packages/composer-runtime/test/engine.logging.js [hunk 7]
```diff
@@ -73,9 +82,9 @@ describe('EngineLogging', () => {
 
         it('should work for the good path', () => {
             mockAccessController.check.resolves();
-            mockLoggingService.getLogLevel.returns('LEVEL');
+
             return engine.invoke(mockContext, 'getLogLevel', ['wrong', 'args', 'count', 'here'])
-            .should.eventually.be.deep.equal('LEVEL');
+            .should.eventually.be.deep.equal('composer[error]:*');
```
Comment (sstone1) [inline packages/composer-runtime/test/engine.logging.js]: This looks suspect; the arguments are clearly wrong, but the test seems to pass?

**Pull request merged**